src/Processors/Executors/ExecutingGraph.cpp

```diff
@@ -263,7 +263,30 @@ bool ExecutingGraph::updateNode(uint64_t pid, Queue & queue, Queue & async_queue
 
                 try
                 {
-                    node.last_processor_status = node.processor->prepare(node.updated_input_ports, node.updated_output_ports);
+                    auto & processor = *node.processor;
+                    IProcessor::Status last_status = node.last_processor_status;
+                    IProcessor::Status status = processor.prepare(node.updated_input_ports, node.updated_output_ports);
+                    node.last_processor_status = status;
+
+                    /// NeedData
+                    if (last_status != IProcessor::Status::NeedData && status == IProcessor::Status::NeedData)
+                    {
+                        processor.need_data_watch.restart();
```

Comment:
Note that similar stopwatch is under define only for DEBUG. I've seen some (a few percent) slow down for some queries because of gettime. We should only measure time if setting is enabled.

Reply:
>Note that similar stopwatch is under define only for DEBUG. I've seen some (a few percent) slow down for some queries because of gettime. 

Interesting, but perf tests does not show this, I guess this is was some synthetic query.
By the way, maybe it was before vdso had been fixed?

>We should only measure time if setting is enabled.

Ok.

Reply:
By the way, I guess one day it will be enabled by default...

Reply:
I did this, and I don't like the complexity that it introduces.

After I've looked at the overhead and I don't see any difference (for `select count() from numbers(1e9)` query):

- patched

<details>

```sh
Queries executed: 100.

localhost:9000, queries 100, QPS: 3.958, RPS: 3958441381.337, MiB/s: 30200.511, result RPS: 3.958, result MiB/s: 0.000.

0.000%          0.238 sec.
10.000%         0.241 sec.
20.000%         0.244 sec.
30.000%         0.246 sec.
40.000%         0.248 sec.
50.000%         0.249 sec.
60.000%         0.252 sec.
70.000%         0.255 sec.
80.000%         0.263 sec.
90.000%         0.269 sec.
95.000%         0.274 sec.
99.000%         0.280 sec.
99.900%         0.285 sec.
99.990%         0.285 sec.
```

</details>

- upstream

<details>

```sh
Queries executed: 100.

localhost:9000, queries 100, QPS: 3.826, RPS: 3825878323.592, MiB/s: 29189.135, result RPS: 3.826, result MiB/s: 0.000.

0.000%          0.244 sec.
10.000%         0.253 sec.
20.000%         0.257 sec.
30.000%         0.258 sec.
40.000%         0.259 sec.
50.000%         0.261 sec.
60.000%         0.262 sec.
70.000%         0.264 sec.
80.000%         0.266 sec.
90.000%         0.270 sec.
95.000%         0.274 sec.
99.000%         0.286 sec.
99.900%         0.287 sec.
99.990%         0.287 sec.
```

</details>

So my suggestion is to add this code only if there will be a proof that it makes the difference.

Reply:
Added in https://github.com/ClickHouse/ClickHouse/pull/34355/commits/5475155f88b94d0be163434be182009e1206b343, but I'm going to remove it if there will be no objections, since I don't like the complexity that it adds.